Clean/Circuit/StructuralLemmas.lean

```diff
@@ -0,0 +1,255 @@
+  have h_circuit2_spec : circuit2.Spec mid (eval env (circuit2.output mid_var (offset + circuit1_len))) := by
+    exact h_holds_circuit2 h_circuit2_assumptions
+
+  -- Use the output equation
+  rw [h_output_eq]
+
+  -- Apply the spec composition
+  apply h_spec_composition input mid _ h_assumptions h_circuit1_spec h_circuit2_spec
+
+
+end Circuit
+
+/--
+Concatenate two FormalCircuits into a single FormalCircuit.
+
+This combinator requires:
+- A compatibility proof that the first circuit's spec implies the second circuit's assumptions
+- A proof that circuit1's output is independent of the offset (h_output_stable)
+
+The composite circuit:
+- Has the assumptions of the first circuit
+- Has a spec stating that there exists an intermediate value such that both component specs hold
+-/
+def FormalCircuit.concat
+    {F : Type} [Field F]
+    {Input Mid Output : TypeMap} [ProvableType Input] [ProvableType Mid] [ProvableType Output]
+    (circuit1 : FormalCircuit F Input Mid)
+    (circuit2 : FormalCircuit F Mid Output)
+    (h_compat : ∀ input mid, circuit1.Assumptions input → circuit1.Spec input mid → circuit2.Assumptions mid)
+    (h_localLength_stable : ∀ mid mid', circuit2.localLength mid = circuit2.localLength mid') :
+    FormalCircuit F Input Output := {
+  elaborated := {
+    main := fun input => circuit1 input >>= circuit2
+    localLength := fun input => circuit1.localLength input + circuit2.localLength (circuit1.output input 0)
+    localLength_eq := by
+      intro input offset
+      simp only [Circuit.bind_def, Circuit.localLength, circuit_norm]
+      -- We need to show that circuit2.localLength at different offsets is the same
+      -- This requires that circuit2.localLength is stable (doesn't depend on its input)
+      congr 1
+      apply h_localLength_stable
+    output := fun input offset =>
+      circuit2.output (circuit1.output input offset) (offset + circuit1.localLength input)
+    output_eq := by
+      intro input offset
+      simp only [Circuit.bind_def, Circuit.output, circuit_norm]
+    subcircuitsConsistent := by
+      intro input offset
+      simp only [Circuit.bind_def, Circuit.operations, circuit_norm]
+      ring
+  }
+  Assumptions := circuit1.Assumptions
+  Spec := fun input output => ∃ mid, circuit1.Spec input mid ∧ circuit2.Spec mid output
+  soundness := by
+    apply Circuit.soundness_compose_circuits (Mid := Mid)
+    · intro; rfl
+    · intro input offset
+      simp only [ElaboratedCircuit.output]
+    · intro input h; exact h
+    · intro input mid h_assumptions h_spec1
+      exact h_compat input mid h_assumptions h_spec1
+    · intro input mid output h_input h_spec1 h_spec2
+      exact ⟨mid, h_spec1, h_spec2⟩
+  completeness := by
+    simp only [circuit_norm]
+    rintro offset env input_var ⟨ use1, use2 ⟩ input h_input asm1
+    constructor
+    · simp only [subcircuit_norm, h_input, asm1]
+    · simp only [subcircuit_norm]
```

Comment:
Both branches are simplifying `subcircuit_norm`. This can be done before `constructor` call.

Reply:
done.